File: 2/LinkedList.java
public class LinkedList {
	Node head = null;

	class Node {
		int data;
		Node next;

		Node(int data) {
			this.data = data;
			this.next = null;
		}
	}

	public void insert(int data) {
		Node cur = this.head;

		if(cur == null) {
			this.head = new Node(data);
			return;
		}
		
		while(cur.next != null) {
			cur = cur.next;
		}
		
		cur.next = new Node(data);
	}

	public void print() {
		Node cur = this.head;

		if(cur != null) {
			while(cur != null) {
				System.out.print(cur.data + " ");
				cur = cur.next;	
			}	
		} 

		System.out.println();
	}
	
	public void deleteFromEnd(int n) {
		if(this.head == null) {
			return;
		}

		Node first = this.head;
		Node second = this.head;

		while(n > 1 && second.next != null) {
			second = second.next;
			n--;
		}
		
		if(second.next == null) {
			this.head = this.head.next;
			return;
		}

		while(second.next.next != null) {
			first = first.next;
			second = second.next;
		}
		
		first.next = first.next.next;
	}

	public static void main(String args[]) {
		LinkedList list = new LinkedList();
		int n = 4;

		for(int i=0; i<=9; i++) {
			list.insert(i);
		}
		
		list.print();
		list.deleteFromEnd(n);
		list.print();
	}
}

File: 2/js/src/LinkedList.ts
class ListNode {
	data: number
	next: ListNode | null

	constructor(data: number) {
		this.data = data;
		this.next = null;
	}
}

class List {
	head: ListNode | null
	tail: ListNode | null

	constructor() {
		this.head = this.tail = null;
	}

	insertEnd(data: number) {
		const newNode = new ListNode(data);

		if(this.head == null) {
			this.head = this.tail = newNode;
			return;
		}

		this.tail.next = newNode;
		this.tail = newNode;
	}

	deleteFromEnd(pos: number) {
		if(this.head == null) {
			return;
		}

		let first, second;
		first = second = this.head;

		while(pos > 0) {
			second = second.next;
			pos--;
		}

		if(second == null) {
			this.head = this.head.next;
			return;
		}

		while(second.next != null) {
			first = first.next;
			second = second.next;
		}

		first.next = first.next.next;
	}

	toString(): string {
		let res = "";
		let cur = this.head;

		while(cur != null) {
			res += cur.data + " ";
			cur = cur.next;
		}

		return res;
	}
}

(function main() {
	const list = new List;

	for(let i=0; i<=9; i++) {
		list.insertEnd(i);
	}

	list.deleteFromEnd(4);

	console.log(list.toString());
})();
